#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [6]:
import pandas as pd

In [8]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [5]:
type(df)

pandas.core.frame.DataFrame

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [9]:
total_gross_income_by_branch = df.groupby('Branch')['gross income'].sum()


In [10]:
total_gross_income_by_branch

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

In [11]:
most_profitable_branch_name = total_gross_income_by_branch.idxmax()
most_profitable_branch_name

'C'

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [19]:
most_profitable_branch_name = total_gross_income_by_branch.idxmax()

In [21]:
branch_city = df[['Branch', 'City']].drop_duplicates()
city_of_top_branch = branch_city[branch_city['Branch'] == most_profitable_branch_name]['City'].values[0]

print(f"Найприбутковіший філіал: {most_profitable_branch_name}")
print(f"Місто, де він розташований: {city_of_top_branch}")

Найприбутковіший філіал: C
Місто, де він розташований: Naypyitaw


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [49]:
invoices_by_category = pd.pivot_table (df, index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc = "count")
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [50]:
top_branch = invoices_by_category['Electronic accessories'].idxmax()
print("Філіал з найбільшим продажем Electronic accessories:", top_branch)

Філіал з найбільшим продажем Electronic accessories: A


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


## Гіпотеза №1

### 📌 Питання:
**Яка категорія товарів є найпопулярнішою серед різних типів клієнтів (`Customer type`)?**

### 🎯 Мета:
Зрозуміти, які товарні категорії (`Product line`) частіше купують **постійні клієнти** та **нові клієнти**. Це допоможе:
- розробити **спеціальні акції** для різних типів клієнтів;
- краще **таргетувати рекламні кампанії**;
- зрозуміти, що цікавіше певній аудиторії.

In [51]:
# Групуємо по Customer type і Product line та рахуємо кількість покупок
popular_by_customer = df.groupby(['Customer type', 'Product line'])['Invoice ID'].count().reset_index()

# Сортуємо для наочності
popular_by_customer = popular_by_customer.sort_values(['Customer type', 'Invoice ID'], ascending=[True, False])

print(popular_by_customer)


   Customer type            Product line  Invoice ID
2         Member      Food and beverages          94
5         Member       Sports and travel          87
1         Member     Fashion accessories          86
4         Member      Home and lifestyle          83
0         Member  Electronic accessories          78
3         Member       Health and beauty          73
6         Normal  Electronic accessories          92
7         Normal     Fashion accessories          92
8         Normal      Food and beverages          80
9         Normal       Health and beauty          79
11        Normal       Sports and travel          79
10        Normal      Home and lifestyle          77


## Висновки за гіпотезою №1:  
### Популярність товарних категорій серед різних типів клієнтів

---

### ✅ Для постійних клієнтів (**Member**):
- Найпопулярніша категорія — **Food and beverages** (94 покупки).
- Інші популярні категорії:
  - **Sports and travel** — 87 покупок
  - **Fashion accessories** — 86 покупок
- Найменш популярна категорія — **Health and beauty** (73 покупки).

**➡️ Висновок:** постійні клієнти віддають перевагу продуктам харчування та товарам для спорту і відпочинку.  
**🎁 Рекомендація:** ефективно пропонувати **акції на продукти та спорттовари**, щоб підвищувати лояльність постійних покупців.

---

### ✅ Для нових клієнтів (**Normal**):
- Дві найпопулярніші категорії:
  - **Electronic accessories** — 92 покупки
  - **Fashion accessories** — 92 покупки
- Далі за популярністю:
  - **Food and beverages** — 80 покупок
- Найменш популярні категорії:
  - **Home and lifestyle** — 77 покупок
  - **Sports and travel** та **Health and beauty** — по 79 покупок.

**➡️ Висновок:** нові клієнти більше цікавляться електронікою та модними товарами.  
**🎁 Рекомендація:** варто робити **рекламні акції на електроніку та аксесуари** для залучення нових клієнтів і стимулювання їхніх перших покупок.

---

### 📌 Загальні рекомендації для бізнесу:
- Для **постійних клієнтів** — пропонувати бонуси на **їжу, спорттовари та товари для дому**.
- Для **нових клієнтів** — акцентувати увагу на **електроніці та модних аксесуарах**.
- **Маркетингові кампанії** можна адаптувати під тип клієнта для підвищення ефективності продажів.


## Гіпотеза №2  
### Чи є різниця в середньому чеку між чоловіками та жінками в різних містах?

---

### 🎯 Практична користь для бізнесу:

- **💡 Краще розуміння поведінки клієнтів:**
    - Аналіз покаже, **хто витрачає більше — чоловіки чи жінки — в різних містах**.
    - Бізнес може точніше сегментувати аудиторію **за статтю та місцем проживання**.

- **🎁 Персоналізація маркетингу:**
    - Наприклад:
        - якщо у місті **Yangon** жінки витрачають більше → **запустити акції для жінок**.
        - якщо у **Mandalay** чоловіки залишають більші суми → **просувати товари для чоловіків**.
    - Це допоможе **підвищити середній чек** та залучити більше покупок.

- **📦 Оптимізація асортименту:**
    - У містах з вищим середнім чеком серед чоловіків можна збільшити пропозицію товарів типу:
        - електроніка, спорт, аксесуари.
    - Для міст з високим середнім чеком серед жінок — зробити акцент на:
        - мода, товари для дому, краса, подорожі.

- **📊 Прогнозування доходів:**
    - Дає можливість **краще планувати виручку по містах і статевих групах клієнтів**, а також **ефективніше розподіляти маркетинговий бюджет**.


In [57]:
avg_check_by_gender_city = df.groupby(['City', 'Gender'])['Total'].mean().reset_index().round(2)


In [58]:
avg_check_by_gender_city

City  Gender   Total
0   Mandalay  Female  326.72
1   Mandalay    Male  313.35
2  Naypyitaw  Female  346.55
3  Naypyitaw    Male  325.89
4     Yangon  Female  330.86
5     Yangon    Male  295.71

## Висновки за гіпотезою №2  
### Середній чек серед чоловіків і жінок у різних містах

---

### ✅ Що ми бачимо з даних:

- **У Mandalay:**
  - Жінки витрачають більше (**326.72**) ніж чоловіки (**313.35**).

- **У Naypyitaw:**
  - Жінки мають найбільший середній чек (**346.55**) серед усіх груп.
  - Чоловіки витрачають трохи менше (**325.89**).

- **У Yangon:**
  - Жінки також витрачають більше (**330.86**) ніж чоловіки (**295.71**).

---

### 📌 Висновки:

- У **всіх містах жінки витрачають більше** ніж чоловіки.
- Найбільше витрачають жінки у місті **Naypyitaw**.
- Чоловіки в місті **Yangon** мають найменший середній чек.

---

### 🎁 Рекомендації для бізнесу:

- У Naypyitaw і Yangon варто більше орієнтуватися на **жіночу аудиторію** — більше акцій, товарів та пропозицій для жінок.
- Для чоловіків у Yangon можна робити **спеціальні акції**, щоб підвищити середній чек.


---
## Гіпотеза №3: Коли найбільше продажів впродовж дня?

Цей аналіз дозволяє визначити **години пікової торгової активності** супермаркету. Розуміння того, коли саме відбувається найбільша кількість продажів, є критично важливим для операційної ефективності.

Ця інформація може бути корисною для:

* **Оптимізації розкладу персоналу:** Забезпечення достатньої кількості касирів та консультантів у години пік, щоб уникнути черг та покращити обслуговування клієнтів.
* **Організації акцій та спеціальних пропозицій:** Запуск маркетингових кампаній або "щасливих годин" саме тоді, коли потік клієнтів найбільший, щоб максимізувати їхню ефективність.
* **Управління запасами:** Забезпечення наявності найпопулярніших товарів на полицях під час підвищеного попиту.

**Висновок:** Аналізуючи дані про продажі за годинами дня, ми можемо виявити найбільш завантажені періоди, що дозволить бізнесу приймати обґрунтовані рішення для підвищення прибутковості та задоволеності клієнтів.

---

In [60]:
df['Time'] = pd.to_datetime(df['Time'], format='%H:%M')
df['Hour_of_Day'] = df['Time'].dt.hour
total_sales_by_hour = df.groupby('Hour_of_Day')['Total'].sum()
peak_sales_hour = total_sales_by_hour.idxmax()
peak_sales_amount = total_sales_by_hour.max()
print(total_sales_by_hour.sort_index())

Hour_of_Day
10    31421.4810
11    30377.3295
12    26065.8825
13    34723.2270
14    30828.3990
15    31179.5085
16    25226.3235
17    24445.2180
18    26030.3400
19    39699.5130
20    22969.5270
Name: Total, dtype: float64


---
---
## Висновки з даних про продажі за годинами дня

З отриманих даних видно, що найбільше продажів зосереджено у певні періоди протягом дня:

* **Ранковий пік (перед обідом):** Є помітний пік близько **10:00 (31421.48)**, а також сильні продажі до обіду.
* **Денний пік (після обіду):** Продажі зростають о **13:00 (34723.23)**, що може свідчити про активність у час обідньої перерви.
* **Вечірній пік (виражений):** Найвищий пік продажів спостерігається о **19:00 (39699.51)**. Це чітко вказує на те, що вечірні години після робочого дня є найактивнішим часом для покупок.
* **Спад продажів:** Продажі помітно знижуються після 19:00, а також є зниження у другій половині дня (16:00-18:00) перед вечірнім піком.

---
## Рекомендації бізнесу

На основі цих висновків можна надати такі рекомендації:

1.  **Оптимізація персоналу у вечірні години:**
    * **Рекомендація:** Максимально збільште кількість персоналу (касирів, консультантів, працівників на викладці товару) з **18:00 до 20:00**. Це дозволить ефективно обслуговувати найвищий потік клієнтів, зменшити черги та покращити клієнтський досвід.
    * **Обґрунтування:** Година 19:00 є абсолютною піковою точкою продажів. Достатня кількість працівників у цей час є ключовою для збереження високої задоволеності клієнтів та максимізації кількості оброблених транзакцій.

2.  **Цілеспрямовані маркетингові акції та промоції:**
    * **Рекомендація:** Розгляньте можливість проведення "щасливих годин" або спеціальних акцій у період з **13:00 до 14:00**, а також акцентуйте рекламні кампанії на період з **17:00 до 19:00**.
    * **Обґрунтування:** Ці години вже показують високу активність. Додаткові стимули у пікові та передпікові години можуть ще більше збільшити обсяги продажів. Можна також спробувати стимулювати продажі у години спаду (наприклад, 16:00-18:00), пропонуючи невеликі знижки або спеціальні пропозиції, щоб вирівняти потік клієнтів.

3.  **Управління запасами та свіжістю товарів:**
    * **Рекомендація:** Забезпечте повне поповнення полиць і наявність свіжих товарів (особливо в категоріях з високим попитом, якщо вони відомі) до **ранкового піку (близько 10:00)**, **обіднього піку (до 13:00)** і обов'язково перед **вечірнім піком (до 18:00)**.
    * **Обґрунтування:** Наявність товару є фундаментальною для здійснення продажу. У пікові години клієнти очікують, що всі необхідні їм товари будуть наявні.

4.  **Аналіз завантаженості в години спаду:**
    * **Рекомендація:** Розгляньте можливість проведення рутинних завдань (наприклад, прийом товару, прибирання, перевірка термінів придатності) у години з меншою активністю клієнтів, наприклад, після **20:00** або у періоди незначного спаду.
    * **Обґрунтування:** Це дозволить персоналу зосередитися на продажах та обслуговуванні клієнтів у години пік, не відволікаючись на інші завдання.

Ці рекомендації, засновані на фактичних даних про продажі, допоможуть супермаркету більш ефективно використовувати свої ресурси та максимізувати прибуток.